### PR TITLE
🐛clusterctl: developer workflow with docker

### PR DESCRIPTION
**Pull request by fabriziopandini**
**What this PR does / why we need it**:
The Docker provider is used for development purposes and it is not going to be published as part of the v1alpha3 packages.

This PR:
1.  deletes the docker provider from the list of available providers in clusterctl (because there will be no official repository). 
2. updates the developer documentation and the local-overrides.py script in order to explain how to continue to use the docker provider for test purposes.

/area clusterctl
/assign @vincepri @wfernandes

### CODE REVIEW

**Comment by ncdc** [comment]
👀

File: docs/book/src/clusterctl/developers.md [hunk 1]
```diff
@@ -77,4 +77,59 @@ please note that each `provider_repo` should have its own `clusterctl-settings.j
...
+
+Run the local-overrides hack.
+
+Edit the clusterctl config file located at `~/.cluster-api/clusterctl.yaml` and configure the docker provider
+by adding the following lines (replace $HOME with your home path):
+
+```bash
+providers:
+  - name: docker
+    url: $HOME/.cluster-api/overrides/infrastructure-docker/latest/infrastructure-components.yaml
```
Comment (ncdc) [inline docs/book/src/clusterctl/developers.md]:
```
$ ./bin/clusterctl init --infrastructure=docker
Fetching providers
Error: failed to get provider components for the "docker" provider: failed to get repository client for the InfrastructureProvider with name docker: error creating the local filesystem repository client: invalid path: path "$HOME/.cluster-api/overrides/infrastructure-docker/latest/infrastructure-components.yaml" must be an absolute path
```

**Comment by ncdc** [comment]
/hold
testing, found some issues, will submit all as a single review

File: docs/book/src/clusterctl/developers.md [hunk 2]
```diff
@@ -77,4 +77,59 @@ please note that each `provider_repo` should have its own `clusterctl-settings.j
...
+  kind create cluster --config ./kind-cluster-with-extramounts.yaml --name clusterapi
+  kubectl cluster-info --context kind-clusterapi
+```
+
+- Run `kind create cluster --config ./kind-cluster-with-extramounts.yaml` to create the management cluster using the above file
+
+- Run `kind load docker-image gcr.io/k8s-staging-capi-docker/capd-manager-amd64:dev` to make the docker provider image available
+  for the kubelet in the management cluster.  
+
+Run clusterctl init
```
Comment (ncdc) [inline docs/book/src/clusterctl/developers.md]: Might want to say that you need to run the command you got from running the local overrides script, as just `clusterctl init` won't suffice, will it?

File: docs/book/src/clusterctl/developers.md (hunk 1 above)
Comment (fabriziopandini) [reply docs/book/src/clusterctl/developers.md]: @ncdc it is required to replace $HOME with your home path (it is described in the comment, might be it is not clear enough...)

Comment (ncdc) [reply docs/book/src/clusterctl/developers.md]: 👍 I just copied the contents. That's on me.

File: cmd/clusterctl/hack/local-overrides.py [hunk 3]
```diff
@@ -127,6 +139,16 @@ def write_local_override(provider, version, components_file, components_yaml):
     except Exception as e:
         raise Exception('failed to write {} to {}: {}'.format(components_file, provider_overrides_folder, e))
 
+def write_docker_metadata(version):
+    try:
+        home = get_home()
+        docker_folder = os.path.join(home, '.cluster-api', 'overrides', 'infrastructure-docker', version)
+        f = open(os.path.join(docker_folder, "metadata.yaml"), 'wb')
```
[the remark this answers is not in the document]
Comment (fabriziopandini) [reply cmd/clusterctl/hack/local-overrides.py]: @ncdc Is there a particolar reason to omit b?
I have got reports from users with python 3.x when storing YAML without b...

[the remark this answers is not in the document]
Comment (ncdc) [reply cmd/clusterctl/hack/local-overrides.py]: It doesn't work with `b`. You get an error that the `f.write` call is expecting bytes, not str.

[the remark this answers is not in the document]
Comment (fabriziopandini) [reply cmd/clusterctl/hack/local-overrides.py]: ...ok
It works on my mac Python 2.7.16

[the remark this answers is not in the document]
Comment (ncdc) [reply cmd/clusterctl/hack/local-overrides.py]: I'm using 3.7

[the remark this answers is not in the document]
Comment (fabriziopandini) [reply cmd/clusterctl/hack/local-overrides.py]: fixed

File: docs/book/src/clusterctl/developers.md (hunk 2 above)
Comment (fabriziopandini) [reply docs/book/src/clusterctl/developers.md]: done!

**Comment by fabriziopandini** [comment]
@ncdc ready for another pass

**Comment by ncdc** [comment]
LGTM, thanks!
/lgtm
/hold cancel

**Pull request merged**